text/0000-target-bundles.md

```diff
@@ -0,0 +1,246 @@
+
+### Target quirks and cfg
+
+Currently we have a few keys dealing with the configuration variables only: `target_os`,
+`target_env`, `target_vendor`, `target_family`, `target_endian`, `target_pointer_width`¹, etc;
+Then there’s options from which some configuration variables are derived, but they are also used to
+tweak compilation: `is_like_windows`, `is_like_osx`, `is_like_solaris`, etc.
+
+This RFC proposes replacing these keys with a different set of keys which explicitly control
+configuration variables and keys which control compilation details:
+
+* `cfg: {"target_env": "msvc", windows: null }` would result in two cofiguration variables
+`#[cfg(windows)]` and `#[cfg(target_env = "msvc")]` evaluating to true, but no variables provided
+by the `cfg` key would get used to influence the behaviour of the compiler; and
+* `debuginfo: ["CodeView", 1]` some targets require different debuginfo format than what LLVM
+generates by default. MSVC targets want CodeView version 1, OS X and Android want Dwarf version 2,
+while LLVM appears to use the highest supported version of Dwarf by default.
+* similarly for many other variables which tweak the way compilation is done.
+
+¹: Technically, `target_pointer_width` is used in trans, but it does not provide any extra
+information over `data_layout`, which is already used elsewhere in the compiler to calculate
+layout, sizing information and alignment of all types, including the pointers.
+
+This would allow selectively reusing various conditional implementations that are present in the
+compiler for a custom target, when none of the `is_like_*` variables would fully suit the target.
+Moreover, being able to specify arbitrary cfg variables would allow easily adapting for various
+miniscule details related to the targets. For example, the targets for ARM CPUs with NEON support
+could export `target_has_neon` without any extra language or compiler support.
+
+That being said, it might make sense to have a whitelist of stable options and cfg variables and
+keep everything else unstable for some duration.
+
+### Proposed JSON key-values
+
+```js
+{
+    // REQUIRED
+    "llvm_target": "x86_64-unknown-linux-gnu",              // LLVM target triple (does not need to match with rustc triple)
+    "data_layout": "e-m:e-i64:64-f80:128-n8:16:32:64-S128", // DataLayout for the target
+    "arch": "x86_64",                                       // Architecture of the target
+
+    // OPTIONAL (should have sane defaults)
+    // Configuration variables injected into the compilation units.
+    "cfg": {
+        "target_os": "linux",
+        "target_family": "unix",
+        "target_arch": "x86_64",
+        "target_endian": "little",
+        "target_pointer_width": "64",
+        "target_env": "gnu",
+        "target_vendor": "unknown",
+        "target_has_atomic": ["8", "16", "32", "64"],       // any of #[cfg(target_has_atomic={"8","16","32","64"}] work.
```

Comment:
@Amanieu previously complained about this being an explicit listing. See https://github.com/rust-lang/rust/pull/38579